deployment/windows/jellyfin.nsi

```diff
@@ -0,0 +1,202 @@
+
+;--------------------------------
+;Pages
+
+    !insertmacro MUI_PAGE_LICENSE "$%InstallLocation%\LICENSE"
+    !insertmacro MUI_PAGE_COMPONENTS
+    !insertmacro MUI_PAGE_DIRECTORY
+
+; Custom Directory page to ask for Emby Library location in case its needed
+; Commented for now to avoid showing this. 
+; This can be uncommented in case Emby Migration is planned later
+;    !define MUI_PAGE_HEADER_TEXT "Emby Library locaton"
+;    !define MUI_PAGE_HEADER_SUBTEXT  ""
+;    !define MUI_DIRECTORYPAGE_TEXT_TOP "Please select the folder where Emby library is present. This will have Enby folders like config, cache, data, metadata, etc."
+;    !define MUI_DIRECTORYPAGE_TEXT_DESTINATION "Emby Library location"
+;    !define MUI_PAGE_CUSTOMFUNCTION_PRE ShowEmbyLibraryPage
+;    !define MUI_DIRECTORYPAGE_VARIABLE $_DEFAULTEMBYDATADIR_
+;    !insertmacro MUI_PAGE_DIRECTORY
+	
+    !insertmacro MUI_PAGE_INSTFILES
+
+    !insertmacro MUI_UNPAGE_CONFIRM
+    !insertmacro MUI_UNPAGE_INSTFILES
+  
+;--------------------------------
+;Languages
+ 
+    !insertmacro MUI_LANGUAGE "English"
+
+;--------------------------------
+;Installer Sections
+
+Section "Install Jellyfin (required)" InstallJellyfin
+    SetOutPath "$INSTDIR"
+    SetShellVarContext current
+    StrCpy $_JELLYFINDATADIR_ "$LOCALAPPDATA\jellyfin\"
+
+; Pack all the files that were just compiled
+    File /r $%InstallLocation%\* 
+	
+; Write the installation path into the registry
+    WriteRegStr HKLM "Software\Jellyfin" "InstallLocation" "$INSTDIR"
+
+; Write the uninstall keys for Windows
+    WriteRegStr HKLM "Software\Microsoft\Windows\CurrentVersion\Uninstall\Jellyfin" "DisplayName" "Jellyfin $_JELLYFINVERSION_"
+    WriteRegStr HKLM "Software\Microsoft\Windows\CurrentVersion\Uninstall\Jellyfin" "UninstallString" '"$INSTDIR\Uninstall.exe"'
+    WriteRegStr HKLM "Software\Microsoft\Windows\CurrentVersion\Uninstall\Jellyfin" "DisplayIcon" '"$INSTDIR\Jellyfin.exe",0'
+    WriteRegStr HKLM "Software\Microsoft\Windows\CurrentVersion\Uninstall\Jellyfin" "Publisher" "The Jellyfin project"
+    WriteRegStr HKLM "Software\Microsoft\Windows\CurrentVersion\Uninstall\Jellyfin" "URLInfoAbout" "https://jellyfin.github.io/"
+    WriteRegStr HKLM "Software\Microsoft\Windows\CurrentVersion\Uninstall\Jellyfin" "DisplayVersion" "$_JELLYFINVERSION_"
+    WriteRegDWORD HKLM "Software\Microsoft\Windows\CurrentVersion\Uninstall\Jellyfin" "NoModify" 1
+    WriteRegDWORD HKLM "Software\Microsoft\Windows\CurrentVersion\Uninstall\Jellyfin" "NoRepair" 1
+
+;Create uninstaller
+    WriteUninstaller "$INSTDIR\Uninstall.exe"
+
+SectionEnd
+
+Section "Jellyfin desktop shortcut" DesktopShortcut
+    SetShellVarContext current
+    DetailPrint "Creating desktop shortcut"
+    CreateShortCut "$DESKTOP\Jellyfin.lnk" "$INSTDIR\jellyfin.exe"
+SectionEnd
+
+Section /o "Jellyfin Service" InstallService
+    ExecWait '"$INSTDIR\nssm.exe" install Jellyfin "$INSTDIR\jellyfin.exe" --datadir "$_JELLYFINDATADIR_"' $0
```

Comment:
This section will have issues if there is already a service named Jellyfin on the system. I am not sure what the best course of action would be here to enable graceful fail or removal of old jellyfin service since I'm not that familiar with NSI. Can we add a step to try uninstall any existing Jellyfin service?

Reply:
Yeah .. so this got interesting. 
I was planning to handle in place upgrades in the next iteration, but put it in now. The latest version handles existing installations and services. Here is a rough pseudo
Existing install & Existing Service - don't ask for folder, just upgrade the files and start service
Existing install but Not service - don't ask for folder, upgrade and give option to install service

The service removal is only handled by uninstall at this stage. 

So there is path to install service while upgrading, but no path to remove service (and go standalone) while upgrading. I believe this is a reasonable expectation. Would love to know your views.

Reply:
Does this approach ensure a clean install directory? Or does it just install over the existing files without removing the old ones?

Reply:
It's an install over the old ones without removing them. 
The path to clean install would be to manually run the uninstaller, which would ensure deletion of the files and then run installer for the new.

Reply:
Would it not be desirable to run the uninstaller as part of the install, if present?
I'd think this would avoid potential problems from files not present in the new version.

Reply:
I'm not sure I understand it completely. Could you elaborate on the ” problems from files not present”.

I'll try to rephrase what I'm saying.
If there is an existing install, we assume the user knows where they want to install and how they want to run (service or standalone). It's a matter of giving them new compiled binaries. So we don't bother them with uninstall and a complete reinstall, just a quick version by skipping pages. I believe it provides a more seamless user experience for someone already using the software.

In case they have an existing install and they haven't realised it, we show a warning at installer start so they can continue or quit. At this stage they can quit and uninstall manually from add/remove programs.

We can do it the way you've mentioned. If an existing install is detected, we can provide the option to run the uninstaller. In my opinion, it breaks the user experience and doesn't fully utilise the installer's capability.
We should also consider what majority of our users are going to follow - are they power users (let me choose everything, everytime) or more like click-click done (I've chosen it once, leave me alone) ?

Reply:
What I mean is that, if a file is for example renamed from one version to the next, this *could* cause hard to track down bugs.

That is, if the installer just overwrites old and creates new files, but not removing old, no longer needed files.

I'm not saying it *will* be a problem, but it *could maybe* be a problem some time in the future.

As for reusing the previously selected options, I'm all for that. I'd love for there to be seamless scripted ("silent") updates.

Reply:
Lots on programs will run the uninstaller silently when in place upgrading. So the user does not even really notice.

Reply:
Thanks @EraYaN @oddstr13 
I understand both the approaches. I have my preferences and I see your point.
To move forward, I can make the changes so now the installer proceeds to uninstall and install transparently.

PS: is there a way to talk outside of GitHub comments for a more "realtime" conversation ?

Reply:
Yes, most of our communication happens on matrix; https://matrix.to/#/+jellyfin:matrix.org